File: android/AmphionRuntime/sdk/src/main/java/com/amphion/asr/internal/ModelLayout.kt
package com.amphion.asr.internal

import com.amphion.asr.ModelType
import java.io.File

/**
 * 把"模型目录 + ModelType"映射到具体的模型文件路径。
 *
 * 协议：每个 model_type 自带一组允许的文件名候选；按顺序找到第一个真实存在的就采用，
 * 找不到就让 EngineImpl 抛 [com.amphion.asr.AsrErrorCode.MODEL_FILE_MISSING]（2002）。
 *
 * 这样 SDK 无需要求所有模型文件名 100% 统一（INT8 / FP32 / fp16 同时存在时优先 INT8），
 * 也避免了硬编码单一文件名（旧实现的局限）。
 */
internal data class ResolvedFiles(
    val encoder: File? = null,
    val decoder: File? = null,
    val joiner: File? = null,
    /** zipformer2_ctc / nemo_ctc 用的单个 model 文件 */
    val model: File? = null,
    val tokens: File,
)

internal object ModelLayout {

    /**
     * 按 model_type 解析模型目录中应当存在的文件，如有缺失返回错误信息。
     *
     * @return Pair(ResolvedFiles?, errorMessage?)。要么 first 非空（成功），要么 second 非空（失败原因）。
     */
    fun resolve(modelDir: File, type: ModelType): Pair<ResolvedFiles?, String?> {
        val tokens = File(modelDir, "tokens.txt")
        if (!tokens.isFile) {
            return null to "missing tokens.txt under ${modelDir.absolutePath}"
        }

        return when (type) {
            ModelType.TRANSDUCER -> {
                val enc = pickFirst(modelDir,
                    "encoder.int8.onnx", "encoder.onnx", "encoder.fp16.onnx")
                val dec = pickFirst(modelDir,
                    "decoder.onnx", "decoder.int8.onnx", "decoder.fp16.onnx")
                val join = pickFirst(modelDir,
                    "joiner.int8.onnx", "joiner.onnx", "joiner.fp16.onnx")
                if (enc == null || dec == null || join == null) {
                    return null to "transducer requires encoder*.onnx + decoder*.onnx + joiner*.onnx under ${modelDir.absolutePath}"
                }
                ResolvedFiles(encoder = enc, decoder = dec, joiner = join, tokens = tokens) to null
            }

            ModelType.PARAFORMER -> {
                val enc = pickFirst(modelDir, "encoder.int8.onnx", "encoder.onnx")
                val dec = pickFirst(modelDir, "decoder.int8.onnx", "decoder.onnx")
                if (enc == null || dec == null) {
                    return null to "paraformer requires encoder*.onnx + decoder*.onnx under ${modelDir.absolutePath}"
                }
                ResolvedFiles(encoder = enc, decoder = dec, tokens = tokens) to null
            }

            ModelType.ZIPFORMER2_CTC, ModelType.NEMO_CTC -> {
                val m = pickFirst(modelDir, "model.int8.onnx", "model.onnx", "model.fp16.onnx")
                    ?: return null to "${type.name.lowercase()} requires model*.onnx under ${modelDir.absolutePath}"
                ResolvedFiles(model = m, tokens = tokens) to null
            }
        }
    }

    private fun pickFirst(dir: File, vararg names: String): File? {
        for (n in names) {
            val f = File(dir, n)
            if (f.isFile) return f
        }
        return null
    }
}

File: asr/ios/Sources/AmphionRuntime/Internal/ModelLayout.swift
import Foundation

/// 与 Android `ModelLayout` 一一对应：按 model_type 解析模型文件路径。
internal struct ResolvedFiles {
    var encoder: URL?
    var decoder: URL?
    var joiner: URL?
    var model: URL?
    let tokens: URL
    var bpeVocab: URL?
}

internal enum ModelLayout {

    static func resolve(modelDir: URL, type: ModelType) throws -> ResolvedFiles {
        let tokens = modelDir.appendingPathComponent("tokens.txt")
        if !FileManager.default.fileExists(atPath: tokens.path) {
            throw AsrError(code: .modelFileMissing, message: "missing tokens.txt under \(modelDir.path)")
        }
        switch type {
        case .transducer:
            let enc = pickFirst(
                modelDir,
                "encoder.int8.ort", "encoder.ort",
                "encoder.int8.onnx", "encoder.onnx", "encoder.fp16.onnx"
            )
            let dec = pickFirst(
                modelDir,
                "decoder.ort", "decoder.int8.ort",
                "decoder.onnx", "decoder.int8.onnx", "decoder.fp16.onnx"
            )
            let join = pickFirst(
                modelDir,
                "joiner.int8.ort", "joiner.ort",
                "joiner.int8.onnx", "joiner.onnx", "joiner.fp16.onnx"
            )
            guard let e = enc, let d = dec, let j = join else {
                throw AsrError(
                    code: .modelFileMissing,
                    message: "transducer requires encoder + decoder + joiner ONNX/ORT files under \(modelDir.path)"
                )
            }
            return ResolvedFiles(
                encoder: e,
                decoder: d,
                joiner: j,
                model: nil,
                tokens: tokens,
                bpeVocab: pickFirst(modelDir, "bbpe.vocab")
            )
        case .paraformer:
            let enc = pickFirst(modelDir, "encoder.int8.ort", "encoder.ort", "encoder.int8.onnx", "encoder.onnx")
            let dec = pickFirst(modelDir, "decoder.int8.ort", "decoder.ort", "decoder.int8.onnx", "decoder.onnx")
            guard let e = enc, let d = dec else {
                throw AsrError(
                    code: .modelFileMissing,
                    message: "paraformer requires encoder + decoder ONNX/ORT files under \(modelDir.path)"
                )
            }
            return ResolvedFiles(
                encoder: e,
                decoder: d,
                joiner: nil,
                model: nil,
                tokens: tokens,
                bpeVocab: nil
            )
        case .zipformer2_ctc, .nemo_ctc:
            guard let m = pickFirst(
                modelDir,
                "model.int8.ort", "model.ort",
                "model.int8.onnx", "model.onnx", "model.fp16.onnx"
            ) else {
                throw AsrError(
                    code: .modelFileMissing,
                    message: "\(type.rawValue) requires model ONNX/ORT file under \(modelDir.path)"
                )
            }
            return ResolvedFiles(
                encoder: nil,
                decoder: nil,
                joiner: nil,
                model: m,
                tokens: tokens,
                bpeVocab: nil
            )
        }
    }

    private static func pickFirst(_ dir: URL, _ names: String...) -> URL? {
        for n in names {
            let f = dir.appendingPathComponent(n)
            if FileManager.default.fileExists(atPath: f.path) { return f }
        }
        return nil
    }
}
